### ISSUE TITLE
GpuWaitResetCollector records zero actions in error

### ISSUE THREAD

**Issue by vuoristo** [astooke/rlpyt#37]
https://github.com/astooke/rlpyt/blob/4dcb17f36d5991082727bc0864fef9ce5054779e/rlpyt/samplers/parallel/gpu/collectors.py#L79

Sometimes the zero action here ends up being treated as a valid action by the code elsewhere. I don't fully understand the parallelization here so I couldn't pinpoint when this failure happens. It can be demonstrated by setting the action to some value you recognize and running example4. In the `MinibatchRl` training loop the values show up for `samples.agent.action[samples.env.done]`.

To me it seems that this line should probably be removed.

**Comment by astooke** [comment]
Hmm, let's see what's happening here...

OK so that line is intended as if the model outputs actions for all environment instances, even the ones which are off (waiting to reset).  So it overwrites those actions to the null value before they are recorded (in the collector) into the batch buffer.  But if you have a non-zero null value, then yes again the best thing would be to write `step.action[b] = env.action_space.null_value()`.

The step buffer has fixed leading dimension of `[B]`, so the model receives step inputs even for environments which are off.  And the batch buffer has fixed leading dimensions of `[T,B]`.

Does that make sense?  Or do you have a different place where you write in the null action values?

**Comment by vuoristo** [comment]
Ok, thanks for the explanation! The buffering logic makes sense to me now. I overlooked the `valid` mask used in the algorithms. The null value itself does not matter too much for me.

However, I still do not understand how things work in the following case. This might be entirely my confusion. If I run the code at the bottom of this comment at commit f0a190c7efe0c55f61ca78dfd30f8d91e9277b24, more often than not the last transition considered valid by the DQN loss has action 0 here: https://github.com/astooke/rlpyt/blob/8331b7f919bcecd3dfe9ece85d2fb479d471ff6f/rlpyt/algos/dqn/dqn.py#L199

I used the following snippet in the DQN loss to check whether there are any zero actions in the loss that are valid according to the mask.

```python
if torch.any(samples.action[valid == 1][samples.done[valid == 1]] == 0):
    import pdb; pdb.set_trace()
```

```python

import gym
import torch

from rlpyt.samplers.parallel.gpu.sampler import GpuSampler
from rlpyt.samplers.parallel.gpu.collectors import GpuWaitResetCollector
from rlpyt.envs.gym import GymEnvWrapper
from rlpyt.algos.dqn.dqn import DQN
from rlpyt.runners.minibatch_rl import MinibatchRl
from rlpyt.agents.dqn.dqn_agent import DqnAgent
from rlpyt.utils.tensor import infer_leading_dims, restore_leading_dims
from rlpyt.models.mlp import MlpModel


NUM_ACTIONS = 100000

class ManyActionDqnModel(torch.nn.Module):

    def __init__(self, **kwargs):
        super().__init__()
        self.head = MlpModel(1, 1, NUM_ACTIONS)

    def forward(self, observation, prev_action, prev_reward):
        lead_dim, T, B, img_shape = infer_leading_dims(observation, 1)
        q = self.head(observation.view(T * B, -1))
        q = restore_leading_dims(q, lead_dim, T, B)
        return q


class ManyActionDqnAgent(DqnAgent):

    def __init__(self, ModelCls=ManyActionDqnModel, **kwargs):
        super().__init__(ModelCls=ModelCls, **kwargs)


class EnvWithManyActions(gym.Env):
    def __init__(self):
        self.observation_space = gym.spaces.Box(0, 1, shape=(1, ))
        self.action_space = gym.spaces.Discrete(NUM_ACTIONS)
        self.steps = 0
    
    def step(self, action):
        self.steps += 1
        return self.observation_space.sample(), 0, self.steps >= 4, {}
    
    def reset(self):
        self.steps = 0
        return self.observation_space.sample()


def make_env(*args, **kwargs):
    return GymEnvWrapper(EnvWithManyActions())


def build_and_train(cuda_idx=None, n_parallel=32):
    sampler = GpuSampler(
        EnvCls=make_env,
        env_kwargs=dict(),
        CollectorCls=GpuWaitResetCollector,
        batch_T=5,
        batch_B=32,
    )
    algo = DQN(batch_size=128)
    agent = ManyActionDqnAgent()
    runner = MinibatchRl(
        algo=algo,
        agent=agent,
        sampler=sampler,
        n_steps=1e8,
        log_interval_steps=1e4,
        affinity=dict(
            cuda_idx=cuda_idx,
            workers_cpus=list(range(n_parallel)),
        ),
    )
    runner.train()

build_and_train()
```

**Comment by astooke** [comment]
Oh!  Good catch on that block under `if not self.mid_batch_reset:` in the dqn loss.  I think that might be wrong.  If not using `mid_batch_reset`, the the replay buffer would have invalid samples which might sometimes be replayed, and I think I was trying to catch this. The collector should record `done=1` for all the time-steps after the first environment `done=1` signal , all the way until the beginning of the next sampler batch, at which point the environment will have been reset.   But without a time-leading dimension, the `valid_from_done()` doesn't work (the first `done` in a sequence is still valid, everything after is invalid).  

Quick fix would be: use a collector _with_ mid-batch resets, which bypasses this block/issue.

Slower fix could be: have the replay buffer check whether each sample is valid--if `done=1`, then make sure the previous time-step has `done=0`, else it's invalid.  And use this instead of the `valid_from_done()` currently in there.


As for the zero actions, I think that is a valid Atari action, corresponding to NOOP (no action)?

**Comment by vuoristo** [comment]
Yeah that makes sense. In my case the `GpuResetCollector` works just fine. Zero seems like a reasonable default action, though it might make the code a bit clearer to store the `null_value` in the collector.

The class doc string in `BaseNStepReturnBuffer` actually documents this issue already but the example4 has the option for using `GpuWaitResetCollector` so I didn't think to look there. https://github.com/astooke/rlpyt/blob/15055b4daf6ab47b51a6fa682a9909f1167fc668/rlpyt/replays/n_step.py#L16

Would it make sense to add the validity computation as a method of the `BaseNStepReturnBuffer` and call it  the same way as `compute_returns`?

**Comment by astooke** [comment]
> The class doc string in `BaseNStepReturnBuffer` actually documents this issue already but the example4 has the option for using `GpuWaitResetCollector` so I didn't think to look there.

OK good point, that example is incomplete, I should at least add a note there.

> Would it make sense to add the validity computation as a method of the `BaseNStepReturnBuffer` and call it the same way as `compute_returns`?

Yes that sounds reasonable, using the same logic mentioned above, if multiple `done=1` steps happen in a row.  I think I debated doing this, but didn't end up putting this into the replay buffer, because I was only using it in the R2D1 algorithm, in which the replay always _starts_ at a valid sample, and returns sequences, on which `valid_from_done()` works.

**Comment by astooke** [comment]
Any news or new questions here? If not, we can close this issue?

**Comment by vuoristo** [comment]
No news at this point. Let's close it.
